skimage/morphology/_util.py

```diff
@@ -3,9 +3,14 @@
 
 import numpy as np
 from scipy import ndimage as ndi
-
-
-def _validate_connectivity(image_dim, connectivity, offset):
+from typing import Union, Optional, Tuple, Sequence
+from ..typing import _SupportsArray
+
+def _validate_connectivity(
+        image_dim: int,
+        connectivity: Optional[Union[int, _SupportsArray]],
+        offset: Optional[Sequence[int]],
+) -> Tuple[np.ndarray, np.ndarray]:
```

Comment:
This might also be something that can be improved even further by using [`NewType`](https://docs.python.org/3/library/typing.html#newtype). The doscstring states that one of the returned arrays is a "structuring element" in the form of an array of bool, and the other is an offset as an array of int.
With `NewType` this can be made explicit and this type can then also be used in the signature of e.g. `_offsets_to_raveled_neighbors` (which is where the return value of this function ends up in in one case)
Its docstring for the parameter states:
```
    selem : ndarray
        A structuring element determining the neighborhood expressed as an
        n-D array of 1's and 0's.
``` 
If `selem` is annotated with the same `NewType` that is returned here, this another example of documentation that can be formalized in the type system for both human and algorithmic use. Unless a `n-D array of 1's and 0's.` is actually different to an `array of bool` which would be even more important to emphasize.

Reply:
This is a good example for the question I had above: how do we maintain parallel hierarchies of (1) NumPy arrays that are NumPy arrays but also a specific type of NumPy arrays, and (2) array-likes that might actually be dask arrays or OpenCL arrays or whatever.